File: fornituras-api/src/main/java/com/numobiz/solutions/fornituras/modules/dashboard/service/DashboardService.java
package com.numobiz.solutions.fornituras.modules.dashboard.service;

import com.numobiz.solutions.fornituras.modules.dashboard.dto.DashboardSummary;
import com.numobiz.solutions.fornituras.modules.equipment.entity.EquipmentStatus;
import com.numobiz.solutions.fornituras.modules.equipment.repository.EquipmentRepository;
import com.numobiz.solutions.fornituras.modules.equipment.repository.EquipmentRepository.StatusTally;
import com.numobiz.solutions.fornituras.modules.equipment.service.ExpiryCalculator;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;

/**
 * Calcula los indicadores del tablero (010) con <b>consultas agregadas</b> del lado servidor: nunca
 * trae el inventario al cliente ni expone PII. Los estados operativos salen de un único
 * {@code GROUP BY status}; la vigencia (próximas a vencer/caducadas) se deriva de
 * {@code fecha_vencimiento} con el mismo criterio de {@link ExpiryCalculator} que usan 001/008, de
 * modo que cada contador coincide con el listado filtrado equivalente (SC-002).
 */
@Service
@Transactional(readOnly = true)
public class DashboardService {

	private final EquipmentRepository equipmentRepository;

	public DashboardService(EquipmentRepository equipmentRepository) {
		this.equipmentRepository = equipmentRepository;
	}

	public DashboardSummary summary() {
		LocalDate today = LocalDate.now();
		Map<EquipmentStatus, Long> byStatus = tallyByStatus();

		long total = byStatus.values().stream().mapToLong(Long::longValue).sum();
		long disponibles = byStatus.getOrDefault(EquipmentStatus.DISPONIBLE, 0L);
		long asignadas = byStatus.getOrDefault(EquipmentStatus.ASIGNADA, 0L);
		long enMantenimiento = byStatus.getOrDefault(EquipmentStatus.EN_MANTENIMIENTO, 0L);

		long caducadas = equipmentRepository
				.countByFechaVencimientoLessThanAndStatusNot(today, EquipmentStatus.BAJA_DEFINITIVA);
		LocalDate warningLimit = today.plusDays(ExpiryCalculator.WARNING_WINDOW_DAYS);
		long proximasAVencer = equipmentRepository
				.countByFechaVencimientoBetweenAndStatusNot(today, warningLimit, EquipmentStatus.BAJA_DEFINITIVA);

		return new DashboardSummary(total, disponibles, asignadas, proximasAVencer, caducadas, enMantenimiento);
	}

	private Map<EquipmentStatus, Long> tallyByStatus() {
		List<StatusTally> tallies = equipmentRepository.tallyByStatus();
		Map<EquipmentStatus, Long> byStatus = new EnumMap<>(EquipmentStatus.class);
		for (StatusTally tally : tallies) {
			byStatus.put(tally.getStatus(), tally.getTotal());
		}
		return byStatus;
	}
}

File: fornituras-api-dotnet/src/Fornituras.Api/Services/DashboardService.cs
using Fornituras.Api.Common;
using Fornituras.Api.Data;
using Fornituras.Api.Dto;
using Microsoft.EntityFrameworkCore;
using EntityEquipmentStatus = Fornituras.Api.Data.Entities.EquipmentStatus;

namespace Fornituras.Api.Services;

public sealed class DashboardService(ApplicationDbContext db) : IDashboardService
{
    public Task<DashboardSummary> GetSummaryAsync(CancellationToken cancellationToken = default) =>
        SummaryAsync(cancellationToken);

    public async Task<DashboardSummary> SummaryAsync(CancellationToken cancellationToken = default)
    {
        var today = DateOnly.FromDateTime(DateTime.UtcNow);
        var warningLimit = today.AddDays(ExpiryCalculator.WarningWindowDays);

        var statusCounts = await db.Equipment.AsNoTracking()
            .GroupBy(e => e.Status)
            .Select(g => new { Status = g.Key, Count = g.LongCount() })
            .ToListAsync(cancellationToken);

        var total = statusCounts.Sum(x => x.Count);
        var disponibles = statusCounts.FirstOrDefault(x => x.Status == EntityEquipmentStatus.DISPONIBLE)?.Count ?? 0;
        var asignadas = statusCounts.FirstOrDefault(x => x.Status == EntityEquipmentStatus.ASIGNADA)?.Count ?? 0;
        var enMantenimiento = statusCounts.FirstOrDefault(x => x.Status == EntityEquipmentStatus.EN_MANTENIMIENTO)?.Count ?? 0;

        var caducadas = await db.Equipment.AsNoTracking()
            .CountAsync(
                e => e.FechaVencimiento != null &&
                     e.FechaVencimiento < today &&
                     e.Status != EntityEquipmentStatus.BAJA_DEFINITIVA,
                cancellationToken);

        var proximas = await db.Equipment.AsNoTracking()
            .CountAsync(
                e => e.FechaVencimiento != null &&
                     e.FechaVencimiento >= today &&
                     e.FechaVencimiento <= warningLimit &&
                     e.Status != EntityEquipmentStatus.BAJA_DEFINITIVA,
                cancellationToken);

        return new DashboardSummary(total, disponibles, asignadas, proximas, caducadas, enMantenimiento);
    }
}
